ivadomed/main.py

```diff
@@ -116,17 +121,33 @@ def cmd_train(context):
     axis_dct = {'sagittal': 0, 'coronal': 1, 'axial': 2}
     # This code will iterate over the folders and load the data, filtering
     # the slices without labels and then concatenating all the datasets together
-    ds_train = loader.BidsDataset(context["bids_path"],
-                                  subject_lst=train_lst,
-                                  target_suffix=context["target_suffix"],
-                                  roi_suffix=context["roi_suffix"],
-                                  contrast_lst=context["contrast_train_validation"],
-                                  metadata_choice=context["metadata"],
-                                  contrast_balance=context["contrast_balance"],
-                                  slice_axis=axis_dct[context["slice_axis"]],
-                                  transform=train_transform,
-                                  multichannel=True if context['multichannel'] else False,
-                                  slice_filter_fn=SliceFilter(**context["slice_filter"]))
+
+    if context["unet_3D"]:
+        ds_train = loader.Bids3DDataset(context["bids_path"],
+                                      subject_lst=train_lst,
+                                      target_suffix=context["target_suffix"],
+                                      roi_suffix=context["roi_suffix"],
+                                      contrast_lst=context["contrast_train_validation"],
+                                      metadata_choice=context["metadata"],
+                                      contrast_balance=context["contrast_balance"],
+                                      slice_axis=axis_dct[context["slice_axis"]],
+                                      transform=train_transform,
+                                      multichannel=context['multichannel'],
+                                      length=context["length_3D"],
+                                      padding=context["padding_3D"])
+    else:
+        ds_train = loader.BidsDataset(context["bids_path"],
+                                      subject_lst=train_lst,
+                                      target_suffix=context["target_suffix"],
+                                      roi_suffix=context["roi_suffix"],
+                                      contrast_lst=context["contrast_train_validation"],
+                                      metadata_choice=context["metadata"],
+                                      contrast_balance=context["contrast_balance"],
+                                      slice_axis=axis_dct[context["slice_axis"]],
+                                      transform=train_transform,
+                                      multichannel=True if context['multichannel'] else False,
+                                      slice_filter_fn=SliceFilter(**context["slice_filter"]))
+
```

Comment:
Almost all parameter are the same between the two loaders, could we have only one class for both and an attribute that could differentiate the two? Or would it be too much work?

Reply:
I put them separate because BidsDataset and Bids3DDataset don't have the same parent class. Do you know a good way to change the parent according to a parameter?

Reply:
Oh! I see..
We could do something like suggested herebelow but not sure if it would be cleaner:
- in `main.py` --> call a function (eg `load_dataset`)with a parameter which controls 2D or 3D
- in `loader.py` --> the function `load_dataset` initialise the BidsDataset with either `Bids3DDataset`, either `BidsDataset`.

I am mainly annoying with that because I am conscious that the `main.py` is getting very long.